test/tools/yulInterpreter/Interpreter.cpp

```diff
@@ -193,29 +197,41 @@ void Interpreter::operator()(Block const& _block)
 			break;
 	}
 
-	closeScope();
+	leaveScope();
 }
 
 u256 Interpreter::evaluate(Expression const& _expression)
 {
-	ExpressionEvaluator ev(m_state, m_dialect, m_variables, m_scopes);
+	ExpressionEvaluator ev(m_state, m_dialect, *m_scope, m_variables);
 	ev.visit(_expression);
 	return ev.value();
 }
 
 vector<u256> Interpreter::evaluateMulti(Expression const& _expression)
 {
-	ExpressionEvaluator ev(m_state, m_dialect, m_variables, m_scopes);
+	ExpressionEvaluator ev(m_state, m_dialect, *m_scope, m_variables);
 	ev.visit(_expression);
 	return ev.values();
 }
 
-void Interpreter::closeScope()
+void Interpreter::enterScope(Block const& _block)
 {
-	for (auto const& [var, funDeclaration]: m_scopes.back())
+	if (!m_scope->subScopes.count(&_block))
+		m_scope->subScopes[&_block] = make_unique<Scope>(Scope{
+			{},
+			{},
+			m_scope
+		});
+	m_scope = m_scope->subScopes[&_block].get();
+}
+
+void Interpreter::leaveScope()
+{
+	for (auto const& [var, funDeclaration]: m_scope->names)
 		if (!funDeclaration)
-			solAssert(m_variables.erase(var) == 1, "");
```

Comment:
We cannot have this check anymore because we might leave a scope earlier than we left in in a previous execution. In that case, the scope might contain more variables than were activated (i.e. it contains more variables than we visited their declarations).